internal/cortex/frontend/transport/handler.go

```diff
@@ -122,14 +145,56 @@ func (f *Handler) ServeHTTP(w http.ResponseWriter, r *http.Request) {
 	r.Body = http.MaxBytesReader(w, r.Body, f.cfg.MaxBodySize)
 	r.Body = io.NopCloser(io.TeeReader(r.Body, &buf))
 
+	if f.cfg.EnableFailedQueryCache {
+		// Store query expression and range length for checks
+		queryExpressionNormalized = NormalizeString(r.URL.Query().Get("query"))
+
+		// Time range length for queries, if either of "start" or "end" are not present, return 0
+		getQueryRangeSeconds := func() int {
+			start, err := strconv.Atoi(r.URL.Query().Get("start"))
+			if err != nil {
+				return 0
+			}
+			end, err := strconv.Atoi(r.URL.Query().Get("end"))
+			if err != nil {
+				return 0
+			}
+			return end - start
+		}
+		queryExpressionRangeLength := getQueryRangeSeconds()
+
+		// Check if query in cache and whether value exceeds time range length
+		if value, ok := f.lru.Get(queryExpressionNormalized); ok && value.(int) >= queryExpressionRangeLength {
+			level.Info(util_log.WithContext(r.Context(), f.log)).Log(
+				"msg", "found expensive query in cache ", "query expression ", queryExpressionNormalized,
+			)
+
+			w.WriteHeader(http.StatusForbidden)
+			level.Error(util_log.WithContext(r.Context(), f.log)).Log("msg", "found query in cache, caused error", "filter for this error log with following query expression ", queryExpressionNormalized)
```

Comment:
Merged two log lines into one error log line.

Reply:
changed to info log due to expected behavior